### PR TITLE
feat: more compact exported snapshot

### PR DESCRIPTION
**Pull request by yihuang**
Closes: #690 #688

- Purely additive changes for iavl library, add new APIs to support a more compact snapshot format.
- Still use the existing `ExportNode` type as the exchange format with sdk, sdk can still use the protobuf encoding as before, which skips empty fields, which works well with the new optimizations.
- Leave the issue of format negotiation and migration to sdk.

### Where to put the code

With the approach taken in this PR, it's totally fine to move the logic to the sdk side as well, if it makes more sense.

### Test Result

Dumped from cronos mainnet data, evm module, version 2000000:

| Snapshot Type             | Size   |
| ------------------------- | ------ |
| Normal snapshot  | 469.4M |
| Compact snapshot | 402M   |
| memiavl snapshot          | 2.3G   |

* The new format further reduce the zlib compressed size by `14%`, not as big as expected, I guess mainly because zlib compression already done a good job.
* Both normal/compact snapshots are dumped with sdk's snapshot manager, with and without the compact format introduced in this PR, sum the size of all the chunks.
* memiavl snapshot is just put for reference, it's a different snapshot format that contains all the node hashes, with zero compression applied.

EDIT:

Just to confirm my suspicion, I recreate the snapshot with zlib compression disabled, it goes like this:

| Snapshot Type    | Size  |
| ---------------- | ----- |
| Normal snapshot  | 1.82G |
| Compact snapshot | 856M  |

The compression rate is `54%`.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,5 +1,9 @@
 # Changelog
 
+### Improvements
+
+- [#]() New APIs `NewCompressExporter`/`NewCompressImporter` to support more compact snapshot format.
```
Comment (yihuang) [inline CHANGELOG.md:5]:
```suggestion
- [#703](https://github.com/cosmos/iavl/pull/703) New APIs `NewCompressExporter`/`NewCompressImporter` to support more compact snapshot format.
```

**Review by tac0turtle — COMMENTED** [review]
lets leave this here for now, in store v2 this could be moved higher, into the sdk

File: compress.go [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+	} else {
+		l = len(b)
+	}
+	for ; off < l; off++ {
+		if a[off] != b[off] {
+			break
+		}
+	}
+	return off
+}
```
Comment (cool-develope) [inline compress.go:137]:
```
func diffOffset(a, b []byte) int {
	var off int
	for ; off < len(a) && l < len(b)l; off++ {
		if a[off] != b[off] {
			break
		}
	}
	return off
}
```

File: compress.go [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+		// apply delta encoding to leaf keys
+		n.Key, e.lastKey = deltaEncode(n.Key, e.lastKey), n.Key
+
+		e.versionStack = append(e.versionStack, n.Version)
+	} else {
+		// branch keys can be derived on the fly when import, safe to skip
+		n.Key = nil
+
+		// delta encode the version
+		maxVersion := maxInt64(e.versionStack[len(e.versionStack)-1], e.versionStack[len(e.versionStack)-2])
```
Comment (cool-develope) [inline compress.go:48]: just curious about the efficiency of version delta encode?

File: compress.go (hunk 2 above)
Comment (cool-develope) [inline compress.go:137]:
```suggestion
func diffOffset(a, b []byte) int {
	var off int
	for ; off < len(a) && off < len(b); off++ {
		if a[off] != b[off] {
			break
		}
	}
	return off
}
```

**Comment by yihuang** [comment]
> just curious about the efficiency of version delta encode?

Just did some more detailed comparisons:

## With Zlib Compression(level 7)

|                                             | Size (bytes) | Compression Ratio |
| ------------------------------------------- | ------------ | ----------------- |
| Original                                    | 492170475    | -                 |
| Skip branch key                             | 455839686    | 7%                |
| Skip branch key + key delta                 | 444661628    | 10%               |
| Skip branch key + key delta + version delta | 421542633    | 14%               |

## Without Zlib Compression

|                                             | Size (bytes) | Compression Ratio |
| ------------------------------------------- | ------------ | ----------------- |
| Original                                    | 1955535429   | -                 |
| Skip branch key                             | 1308448614   | 33%               |
| Skip branch key + key delta                 | 936722517    | 52%               |
| Skip branch key + key delta + version delta | 897597115    | 54%               |

File: compress.go (hunk 2 above)
Comment (yihuang) [reply compress.go:137]: This change will do two comparisons instead of one in the for loop

File: compress.go (hunk 3 above)
Comment (yihuang) [reply compress.go:48]: https://github.com/cosmos/iavl/pull/703#issuecomment-1472958077

**Comment by cool-develope** [comment]
txs for detail report

**Review by cool-develope — APPROVED** [review]

**Comment by tac0turtle** [comment]
@mergifyio backport release/v1.x.x

**Pull request merged**